pkg/gateway/operations/deleteobjects.go

```diff
@@ -56,43 +69,107 @@ func (controller *DeleteObjects) Handle(w http.ResponseWriter, req *http.Request
 				Key:     obj.Key,
 				Message: "Access Denied",
 			})
-		}
-
-		lg := o.Log(req).WithField("key", obj.Key)
-		err = o.Catalog.DeleteEntry(req.Context(), o.Repository.Name, resolvedPath.Ref, resolvedPath.Path)
-		switch {
-		case errors.Is(err, catalog.ErrNotFound),
-			errors.Is(err, graveler.ErrNotFound):
-			lg.Debug("tried to delete a non-existent object (OK)")
-		case errors.Is(err, graveler.ErrWriteToProtectedBranch):
-			_ = o.EncodeError(w, req, gerrors.Codes.ToAPIErr(gerrors.ErrWriteToProtectedBranch))
-		case errors.Is(err, catalog.ErrPathRequiredValue):
-			// issue #1706 - https://github.com/treeverse/lakeFS/issues/1706
-			// Spark trying to delete the path "main/", which we map to branch "main" with an empty path.
-			// Spark expects it to succeed (not deleting anything is a success), instead of returning an error.
-			lg.Debug("tried to delete with an empty branch")
-		case err != nil:
-			lg.WithError(err).Error("failed deleting object")
-			errs = append(errs, serde.DeleteError{
-				Code:    "ErrDeletingKey",
-				Key:     obj.Key,
-				Message: fmt.Sprintf("error deleting object: %s", err),
-			})
 			continue
-		default:
-			lg.Debug("object set for deletion")
 		}
-		if !decodedXML.Quiet {
-			responses = append(responses, serde.Deleted{Key: obj.Key})
+
+		keysToDelete = append(keysToDelete, obj.Key)
+		refsToDelete = append(refsToDelete, resolvedPath.Ref)
+		pathsToDelete = append(pathsToDelete, resolvedPath.Path)
+	}
+	if len(pathsToDelete) == 0 {
+		// construct response
+		resp := serde.DeleteResult{Deleted: responses, Error: errs}
+		o.EncodeResponse(w, req, resp, http.StatusOK)
+		return
+	}
+
+	// batch delete - if all paths to delete are same ref
+	canBatch := true
+	for i := 1; i < len(refsToDelete); i++ {
+		if refsToDelete[0] != refsToDelete[i] {
+			canBatch = false
+			break
 		}
 	}
-	// construct response
-	resp := serde.DeleteResult{}
+
+	var resp serde.DeleteResult
+	if canBatch {
+		// batch - call batch delete and translate error/responses
+		resp = controller.batchDelete(req.Context(), o.Log(req), o, decodedXML.Quiet, refsToDelete[0], keysToDelete, pathsToDelete)
+	} else {
+		// non batch - call delete for each key
+		resp = controller.nonBatchDelete(req.Context(), o.Log(req), o, decodedXML.Quiet, keysToDelete, refsToDelete, pathsToDelete)
+	}
+
+	// construct response - concat what we had so far with delete results
 	if len(errs) > 0 {
-		resp.Error = errs
+		resp.Error = append(errs, resp.Error...)
 	}
-	if !decodedXML.Quiet && len(responses) > 0 {
-		resp.Deleted = responses
+	if len(responses) > 0 {
```

Comment:
Where is responses being used? I'm confused

Reply:
Fixed - leftovers from the refactor.